maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/DistributionStageImpl.java

```diff
@@ -101,34 +83,100 @@ public boolean accept(File file) {
+        toExtract.mkdirs();
+        String downloadedPath = downloaded.getAbsolutePath();
+        extractFile(downloaded, toExtract, downloadedPath);
+        return toExtract;
+    }
 
-                } else if (downloadedPath.endsWith(".tar.gz")) {
-                    Spacelift.task(downloaded, UntarTool.class).gzip(true).toDir(toExtract).execute().await();
+    private void extractFile(File downloaded, File toExtract, String downloadedPath) {
+        File tempFile = createTempFile(toExtract);
+        try {
+            if (downloadedPath.endsWith(".zip")) {
+                Spacelift.task(downloaded, UnzipTool.class).toDir(toExtract).execute().await();
+            } else if (downloadedPath.endsWith(".tar.gz")) {
+                Spacelift.task(downloaded, UntarTool.class).gzip(true).toDir(toExtract).execute().await();
+            } else if (downloadedPath.endsWith(".tar.bz2")) {
+                Spacelift.task(downloaded, UntarTool.class).bzip2(true).toDir(toExtract).execute().await();
+            } else {
+                throw new IllegalArgumentException(
+                   "The distribution " + downloaded + " is compressed by unsupported format. "
+                      + "Supported formats are .zip, .tar.gz, .tar.bz2");
+            }
+        } catch (ExecutionException ee) {
+            throw new IllegalStateException(
+               "Something bad happened when the file: " + downloadedPath + " was being extracted. "
+                  + "For more information see the stacktrace", ee);
+        }
+        if (!tempFile.delete()) {
+            log.warning("failed to delete temp directory: " + tempFile);
+        }
+        System.out.println(String.format("Resolver: Successfully extracted maven binaries from %s", downloaded));
+    }
 
-                } else if (downloadedPath.endsWith(".tar.bz2")) {
-                    Spacelift.task(downloaded, UntarTool.class).bzip2(true).toDir(toExtract).execute().await();
+    private File checkIfItIsAlreadyExtracted(final String downloadedZipMd5hash) {
+
+        File[] dirs = new File(MAVEN_TARGET_DIR).listFiles(new FileFilter() {
 
-                } else {
-                    throw new IllegalArgumentException(
-                            "The distribution " + downloaded + " is compressed by unsupported format. "
-                                    + "Supported formats are .zip, .tar.gz, .tar.bz2");
+            @Override
+            public boolean accept(File file) {
+                return file.isDirectory() && downloadedZipMd5hash.equals(file.getName());
+            }
+        });
+        if (dirs != null && isExtractionFinished(dirs)) {
+            return dirs[0];
+        }
+        return null;
+    }
+
+    private boolean isExtractionFinished(File[] dirs) {
+        if (dirs.length > 0) {
+            File dir = dirs[0];
+            int count = 0;
+            while (isTempFilePresent(dir) && count < 10) {
```

Comment:
if I count it correctly, the timeout is one second...? I would raise it at least to 5 or 10 seconds.